src/lib/Lower.hs

```diff
@@ -112,6 +119,216 @@ traverseExprWithDest dest expr = case expr of
+      _ -> emitDecl (getNameHint b) ann =<< applySubst frag expr
+    vectorizeLoopsRec (frag <.> b @> v) rest
+
+vectorizeSeq :: forall i i' o. (Distinct o, Ext i o)
+             => Word32 -> Type o -> SubstFrag Name i i' o -> LamExpr i'
+             -> TopVectorizeM o (Atom o)
+vectorizeSeq loopWidth newIxTy frag (LamExpr (LamBinder b ty arr eff) body) = do
+  case eff of
+    Pure -> do
+      ty' <- case ty of
+        ProdTy [_, ref] -> do
+          ref' <- applySubst frag ref
+          return $ ProdTy [newIxTy, ref']
+        _ -> error "Unexpected seq binder type"
+      liftM (runHardFail . (`runReaderT` loopWidth)) $ liftBuilderT $ do
+        buildLam (getNameHint b) arr (sink ty') Pure \ci -> do
+          (vi, dest) <- fromPair $ Var ci
+          viOrd <- emitOp $ CastOp IdxRepTy vi
+          iOrd <- emitOp $ BinOp IMul viOrd (IdxRepVal loopWidth)
+          i <- emitOp $ CastOp (sink newIxTy) iOrd
+          let s = newSubst iSubst <>> b @> VVal (ProdStability [Contiguous, Uniform]) (PairVal i dest)
+          runSubstReaderT s $ runVectorizeM $ vectorizeBlock body $> UnitVal
+    _ -> error "Effectful loop vectorization not implemented!"
+  where
+    iSubst :: (DExt o o', Color c) => Name c i' -> VSubstValC c o'
+    iSubst v = case lookupSubstFragProjected frag v of
+      Left v'  -> sink $ fromNameVAtom v'
+      Right v' -> sink $ fromNameVAtom v'
+
+fromNameVAtom :: forall c n. Color c => Name c n -> VSubstValC c n
+fromNameVAtom v = case eqColorRep @c @AtomNameC of
+  Just ColorsEqual -> VVal Uniform $ Var v
+  _ -> error "Unexpected non-atom name"
+
+newtype VectorizeM i o a =
+  VectorizeM { runVectorizeM :: SubstReaderT VSubstValC (BuilderT (ReaderT Word32 HardFailM)) i o a }
+  deriving ( Functor, Applicative, Monad, Fallible, MonadFail, SubstReader VSubstValC
+           , Builder, EnvReader, EnvExtender, ScopeReader, ScopableBuilder )
+
+getLoopWidth :: VectorizeM i o Word32
+getLoopWidth = VectorizeM $ SubstReaderT $ ReaderT $ const $ ask
+
+vectorizeBlock :: Emits o => Block i -> VectorizeM i o (VAtom o)
+vectorizeBlock (Block _ decls topAns) = go topAns decls
+  where
+    go :: Emits o => Atom i' -> Nest Decl i i' -> VectorizeM i o (VAtom o)
+    go ans = \case
+      Empty -> vectorizeAtom ans
+      Nest (Let b (DeclBinding ann _ expr)) rest -> do
+        unless (ann == PlainLet) $ error "Encountered a non-plain let?"
+        v <- vectorizeExpr expr
+        extendSubst (b @> v) $ go ans rest
+
+vectorizeExpr :: Emits o => Expr i -> VectorizeM i o (VAtom o)
+vectorizeExpr expr = case expr of
+  Atom atom -> vectorizeAtom atom
+  Op op -> vectorizeOp op
+  Hof (RunIO (Lam (LamExpr (LamBinder b UnitTy PlainArrow _) body))) -> do
+    -- TODO: buildBlockAux?
+    (vy, lam') <- withFreshBinder (getNameHint b) UnitTy \b' -> do
+      Abs decls (LiftE vy `PairE` yWithTy) <- buildScoped do
+        VVal vy y <- extendSubst (b @> VVal Uniform UnitVal) $ vectorizeBlock body
```

Comment:
Yes, that's a great observation. I think there are classes of IO computations that are safe to vectorize (e.g. loads) and those that might not be (e.g. FFI). I'm pretty sure that at the moment we're good, because the only IO-inducing op we can vectorize is `LoadOp` and anything else would cause the vectorization pass to fail. In some sense, we fuse the safety checks into vectorization itself, because if the body would not be safe to vectorize, then it would not understand the ops anyway.